File: 33-SearchInRotatedSortedArray/33-SearchInRotatedSortedArray.py
# Last updated: 7/2/2025, 11:55:26 PM
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        n = len(nums)        
        l, r = 0, n-1
      
        while(l<=r):
            mid = (l + r) // 2

            if nums[mid] == target:
                return mid
            if nums[l] <= nums[mid]:
                if nums[l] <= target <= nums[mid]:
                    r = mid
                else:
                    l = mid+1
            
            # if it lies in second half
            else:
                if nums[mid] <= target <= nums[r]:
                    l = mid
                else:
                    r = mid-1
            
            
        return -1
        

File: 33-SearchInRotatedSortedArray/33-SearchInRotatedSortedArray.cpp
// Last updated: 7/2/2025, 11:55:22 PM
class Solution {
public:
    int search(vector<int>& arr, int x) {
        // we need to check the sorting of it,
        // if not the end might be small and first will be big"
        // [7 , 8 , 9 ,0, 1, 2,3,4]

        int l = 0, h = arr.size() - 1;

        while(l <= h)
        {
          int mid = l + (h-l)/2;

          if(arr[mid] == x) return mid;
          // left side sorted
          if(arr[l] <= arr[mid])
          {
              if(arr[l] <= x && x <= arr[mid])
                h = mid-1;
              else l = mid+1;
          }
          else{
            
              if(arr[mid] <= x && x <= arr[h])
                l = mid+1;
              else h = mid-1;
          
          }
        }

      return -1;
    }
};